src/box/iproto.cc

```diff
@@ -1314,7 +1325,14 @@ iproto_flush(struct iproto_connection *con)
 		begin->pos += advance;
 		assert(begin->pos <= end->pos);
 	} else if (nwr < 0 && ! sio_wouldblock(errno)) {
-		diag_raise();
+		/*
+		 * Don't close the connection on write error. Log the error and
+		 * don't write to the socket anymore. Continue processing
+		 * requests as usual, because the client might have closed the
+		 * socket, but still expect pending requests to complete.
+		*/
+		diag_log();
+		con->can_write = false;
 	}
 	return -1;
```

Comment:
One more thing: this function is expected to return -1 only if the socket is blocked. You should set *begin to *end here and return 0 in case of socket write error. Please also move `return -1` upper, to the `if(){}` block where it belongs:

```c
if (nwr > 0) {
    /* ... */
    return -1;
} else if (nwr < 0 && ! sio_wouldblock(errno)) {
    /* ... */
    *begin = *end;
    return 0;
}
```

Reply:
Yes i see this, may be `goto` in if(!con->can_write)?

Reply:
`*begin = *end` is a self-explaining statement (resets the buffers) and it occupies just one line of code so using a goto here would make the code neither easier for understanding nor shorter, hence it's not necessary. IMO in general it's better to avoid using goto except for error reporting, because it breaks the code flow, thus making the code more difficult for understanding.

Reply:
fixed